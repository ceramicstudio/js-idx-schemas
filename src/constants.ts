import type { IDXPublishedDefinitions, IDXPublishedSchemas } from './types'

export const publishedDefinitions: IDXPublishedDefinitions = {
  basicProfile: 'kjzl6cwe1jw149gzi9cwo5ew3w7zilgnkkpbc3ilz2vpszx0dspwehiofogpis1',
  cryptoAccounts: 'kjzl6cwe1jw148845sqymlnkea5dll9p7om78lcndtjhvj5y9u3qokkrhan3imr',
  threeIdKeychain: 'kjzl6cwe1jw146of0djwju60si5cmbadn98ztk3mootwwekhvmusyvwlmom0nrr',
}

export const publishedSchemas: IDXPublishedSchemas = {
  BasicProfile: 'ceramic://k3y52l7qbv1frxvsxw16w4zxgbuji4yk74xlfub711wrfjvaugf2x0sg278ameghs',
  CryptoAccounts: 'ceramic://k3y52l7qbv1fry4ai1jnmjqo6260zo3z0u5iyxvx2xpaukon6o0x39rkkbjl0dqf4',
  Definition: 'ceramic://k3y52l7qbv1fry6zfns4ik8xpl4gfuqinac50b17j3z0f32vgie8v9rt235cv72f4',
  IdentityIndex: 'ceramic://k3y52l7qbv1frxz6n89jr8m3dqz7i2179m3wtg21v63gnxu1oo3xr8yle1xabw9a8',
  ThreeIdKeychain: 'ceramic://k3y52l7qbv1frymaxbddcyi0usku9px485z87s2i9v2h6cuofmhmb9pflizzfttds',
}
