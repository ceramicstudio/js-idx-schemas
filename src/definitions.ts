import type { DID } from 'dids'

import { signIDXDefinitions } from './signing'
import type {
  Definition,
  IDXDefinitionName,
  IDXPublishedSchemas,
  IDXSignedDefinitions,
} from './types'

export function createIDXDefinitions(
  schemas: IDXPublishedSchemas
): Record<IDXDefinitionName, Definition> {
  return {
    basicProfile: {
      name: 'Basic Profile',
      schema: schemas.BasicProfile,
    },
    cryptoAccountLinks: {
      name: 'Crypto Account Links',
      schema: schemas.CryptoAccountLinks,
    },
    threeIdKeychain: {
      name: '3ID Keychain',
      schema: schemas.ThreeIdKeychain,
    },
  }
}

export async function createIDXSignedDefinitions(
  did: DID,
  schemas: IDXPublishedSchemas
): Promise<IDXSignedDefinitions> {
  const definitions = createIDXDefinitions(schemas)
  return await signIDXDefinitions(did, schemas.Definition, definitions)
}
