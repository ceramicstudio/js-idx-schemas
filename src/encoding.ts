import CID from 'cids'
import type { DagJWS, DagJWSResult } from 'dids'

import type { EncodedDagJWS, EncodedDagJWSResult } from './types'
import { applyMap } from './utils'

export function decodeDagJWS({ payload, signatures, link }: EncodedDagJWS): DagJWS {
  return { payload, signatures, link: new CID(link) }
}

export function encodeDagJWS({ payload, signatures, link }: DagJWS): EncodedDagJWS {
  // eslint-disable-next-line
  return { payload, signatures, link: link.toString() }
}

export function decodeDagJWSResult({ jws, linkedBlock }: EncodedDagJWSResult): DagJWSResult {
  // eslint-disable-next-line
  return { jws: decodeDagJWS(jws), linkedBlock: Buffer.from(linkedBlock, 'base64') }
}

export function encodeDagJWSResult({ jws, linkedBlock }: DagJWSResult): EncodedDagJWSResult {
  return { jws: encodeDagJWS(jws), linkedBlock: Buffer.from(linkedBlock).toString('base64') }
}

export function decodeSignedMap<K extends string>(
  data: Record<K, EncodedDagJWSResult>
): Record<K, DagJWSResult> {
  return applyMap(data, decodeDagJWSResult)
}

export function encodeSignedMap<K extends string>(
  data: Record<K, DagJWSResult>
): Record<K, EncodedDagJWSResult> {
  return applyMap(data, encodeDagJWSResult)
}
