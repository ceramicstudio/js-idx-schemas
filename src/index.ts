import { CeramicApi } from '@ceramicnetwork/ceramic-common'

import { isSchemaSecure } from './validate'

export * as allSchemas from './schemas'
export { isSchemaSecure } from './validate'

export interface SchemaItem {
  docId?: string
  name: string
  schema: Record<string, any>
}

export async function publishSchema(ceramic: CeramicApi, item: SchemaItem): Promise<string> {
  if (!isSchemaSecure(item.schema)) {
    throw new Error(`Schema ${item.name} is not secure`)
  }

  if (item.docId == null) {
    const doc = await ceramic.createDocument('tile', { content: item.schema })
    return doc.id
  }

  const doc = await ceramic.loadDocument(item.docId)
  await doc.change({ content: item.schema })
  return doc.id
}

export interface PublishConfig {
  ceramic: CeramicApi
  schemas: Array<SchemaItem>
}

export async function publishSchemas({
  ceramic,
  schemas,
}: PublishConfig): Promise<Record<string, string>> {
  const docIds = await Promise.all(
    schemas.map(async (item: SchemaItem) => await publishSchema(ceramic, item))
  )
  return schemas.reduce((acc, { name }, i) => {
    acc[name] = docIds[i]
    return acc
  }, {} as Record<string, string>)
}
