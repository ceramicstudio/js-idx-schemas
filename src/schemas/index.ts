import BasicProfile from './BasicProfile.json'
import DocIdDocIdMap from './DocIdDocIdMap.json'
import DocIdMap from './DocIdMap.json'
import StringMap from './StringMap.json'

export { BasicProfile, DocIdDocIdMap, DocIdMap, StringMap }
