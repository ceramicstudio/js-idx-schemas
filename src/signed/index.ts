import { decodeDagJWSResult, decodeSignedMap } from '../encoding'

import encodedDefinitions from './definitions.json'
import encodedDID from './did.json'
import encodedSchemas from './schemas.json'

export const signedDefinitions = decodeSignedMap(encodedDefinitions)
export const signedDID = encodedDID.map(decodeDagJWSResult)
export const signedSchemas = decodeSignedMap(encodedSchemas)
