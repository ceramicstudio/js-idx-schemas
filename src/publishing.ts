import type { CeramicApi, DocMetadata, DocOpts, Doctype } from '@ceramicnetwork/common'
import { schemas as publishedSchemas } from '@ceramicstudio/idx-constants'
import type {
  IDXDefinitionName,
  IDXPublishedDefinitions,
  IDXPublishedSchemas,
  IDXSchemaName,
} from '@ceramicstudio/idx-constants'
import type { DagJWSResult } from 'dids'
import isEqual from 'fast-deep-equal'

import { signedDefinitions, signedSchemas } from './signed'
import type {
  Definition,
  DefinitionDoc,
  IDXPublishedConfig,
  IDXSignedDefinitions,
  IDXSignedSchemas,
  PublishDoc,
  SchemaDoc,
} from './types'
import { promiseMap, docIDToString } from './utils'
import { isValidDefinition, isSecureSchema } from './validate'

const PUBLISH_OPTS: DocOpts = { anchor: false, publish: false }

export async function createTile<T = unknown>(
  ceramic: CeramicApi,
  content: T,
  metadata: Partial<DocMetadata> = {}
): Promise<Doctype> {
  if (ceramic.did == null) {
    throw new Error('Ceramic instance is not authenticated')
  }

  if (metadata.controllers == null || metadata.controllers.length === 0) {
    metadata.controllers = [ceramic.did.id]
  }

  const doc = await ceramic.createDocument('tile', { content, metadata: metadata as DocMetadata })
  await ceramic.pin.add(doc.id)
  return doc
}

export async function publishDoc<T = unknown>(
  ceramic: CeramicApi,
  doc: PublishDoc<T>
): Promise<Doctype> {
  if (doc.id == null) {
    return await createTile(ceramic, doc.content, {
      controllers: doc.controllers,
      schema: doc.schema ? docIDToString(doc.schema) : undefined,
    })
  }

  const loaded = await ceramic.loadDocument(doc.id)
  if (!isEqual(loaded.content, doc.content)) {
    await loaded.change({ content: doc.content })
  }
  return loaded
}

export async function createDefinition(
  ceramic: CeramicApi,
  definition: Definition
): Promise<Doctype> {
  if (!isValidDefinition(definition)) {
    throw new Error('Invalid definition')
  }
  return await createTile(ceramic, definition, { schema: publishedSchemas.Definition })
}

export async function updateDefinition(ceramic: CeramicApi, doc: DefinitionDoc): Promise<boolean> {
  const loaded = await ceramic.loadDocument(doc.id)
  if (loaded.metadata.schema !== publishedSchemas.Definition) {
    throw new Error('Document is not a valid Definition')
  }

  if (!isEqual(loaded.content, doc.content)) {
    await loaded.change({ content: doc.content })
    return true
  }
  return false
}

export async function publishRecords(
  ceramic: CeramicApi,
  [genesis, ...updates]: Array<DagJWSResult>
): Promise<Doctype> {
  const doc = await ceramic.createDocumentFromGenesis('tile', genesis, PUBLISH_OPTS)
  await ceramic.pin.add(doc.id)
  for (const record of updates) {
    await ceramic.applyRecord(doc.id, record, PUBLISH_OPTS)
  }
  return doc
}

export async function publishSchema(ceramic: CeramicApi, doc: SchemaDoc): Promise<Doctype> {
  if (!isSecureSchema(doc.content)) {
    throw new Error(`Schema ${doc.name} is not secure`)
  }
  return await publishDoc(ceramic, doc)
}

export async function publishSignedMap<T extends string = string>(
  ceramic: CeramicApi,
  signed: Record<T, Array<DagJWSResult>>
): Promise<Record<T, Doctype>> {
  return await promiseMap(signed, async (records) => await publishRecords(ceramic, records))
}

export async function publishIDXSignedDefinitions(
  ceramic: CeramicApi,
  definitions: IDXSignedDefinitions = signedDefinitions
): Promise<IDXPublishedDefinitions> {
  const signedMap = await publishSignedMap(ceramic, definitions)
  return Object.entries(signedMap).reduce((acc, [key, doc]) => {
    acc[key as IDXDefinitionName] = doc.id.toString()
    return acc
  }, {} as IDXPublishedDefinitions)
}

export async function publishIDXSignedSchemas(
  ceramic: CeramicApi,
  schemas: IDXSignedSchemas = signedSchemas
): Promise<IDXPublishedSchemas> {
  const signedMap = await publishSignedMap(ceramic, schemas)
  return Object.entries(signedMap).reduce((acc, [key, doc]) => {
    acc[key as IDXSchemaName] = doc.versionId.toUrl()
    return acc
  }, {} as IDXPublishedSchemas)
}

export async function publishIDXConfig(ceramic: CeramicApi): Promise<IDXPublishedConfig> {
  const [definitions, schemas] = await Promise.all([
    publishIDXSignedDefinitions(ceramic),
    publishIDXSignedSchemas(ceramic),
  ])
  return { definitions, schemas }
}
