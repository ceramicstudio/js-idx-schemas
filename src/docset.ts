import type { CeramicApi, DocMetadata, Doctype } from '@ceramicnetwork/common'
import type DocID from '@ceramicnetwork/docid'
import type { DagJWSResult } from 'dids'

import { decodeSignedMap, encodeSignedMap } from './encoding'
import { createDefinition, createTile, publishRecords, publishSchema } from './publishing'
import type { Definition, EncodedDagJWSResult, Schema } from './types'
import { docIDToString } from './utils'

export type CreatedDoc = {
  id: DocID
  dependencies: Array<DocID>
}

export type PublishedDocSet = {
  definitions: Record<string, string>
  schemas: Record<string, string>
  tiles: Record<string, string>
}

export type DocSetData<T> = {
  docs: Record<string, T>
  definitions: Array<string>
  schemas: Array<string>
}
export type SignedDocSet = DocSetData<Array<DagJWSResult>>
export type EncodedSignedDocSet = DocSetData<Array<EncodedDagJWSResult>>

export class DocSet {
  _ceramic: CeramicApi
  _docs: Record<string, Promise<Doctype>> = {}
  _definitions: Record<string, Promise<CreatedDoc>> = {}
  _schemas: Record<string, Promise<CreatedDoc>> = {}
  _schemaAliases: Record<string, string> = {}
  _tiles: Record<string, Promise<CreatedDoc>> = {}

  constructor(ceramic: CeramicApi) {
    if (ceramic.did == null) {
      throw new Error('Ceramic instance must be authenticated')
    }
    this._ceramic = ceramic
  }

  async _loadDoc(docID: DocID | string): Promise<Doctype> {
    const id = docIDToString(docID)
    if (this._docs[id] == null) {
      this._docs[id] = this._ceramic.loadDocument(id)
    }
    return await this._docs[id]
  }

  get definitions(): Array<string> {
    return Object.keys(this._definitions)
  }

  get schemas(): Array<string> {
    return Object.keys(this._schemas)
  }

  get tiles(): Array<string> {
    return Object.keys(this._tiles)
  }

  hasSchema(alias: string): boolean {
    return this._schemas[alias] != null
  }

  deleteSchema(alias: string): boolean {
    if (this.hasSchema(alias)) {
      delete this._schemas[alias]
      return true
    }
    return false
  }

  createSchema(
    name: string,
    schema: Schema,
    deps: Array<Promise<DocID>> = []
  ): Promise<CreatedDoc> {
    if (this.hasSchema(name)) {
      throw new Error(`Schema ${name} already exists`)
    }

    this._schemas[name] = Promise.all(deps).then((dependencies) => {
      return publishSchema(this._ceramic, { name, content: schema }).then(
        (doc) => {
          this._schemaAliases[doc.versionId.toUrl()] = name
          return { id: doc.versionId, dependencies }
        },
        (reason: any) => {
          delete this._schemas[name]
          throw reason
        }
      )
    })
    return this._schemas[name]
  }

  async addSchema(schema: Schema, alias?: string): Promise<DocID> {
    const name = alias ?? (schema.title as string | undefined)
    if (name == null) {
      throw new Error('Schema must have a title property or an alias must be provided')
    }

    // TODO: lookup CeramicSchemaRef in the schema definitions
    // If set, add schemas as dependencies

    const created = await this.createSchema(name, schema)
    return created.id
  }

  hasDefinition(alias: string): boolean {
    return this._definitions[alias] != null
  }

  deleteDefinition(alias: string): boolean {
    if (this.hasDefinition(alias)) {
      delete this._definitions[alias]
      return true
    }
    return false
  }

  createDefinition(
    alias: string,
    definition: Definition,
    deps: Array<Promise<DocID>> = []
  ): Promise<CreatedDoc> {
    if (this.hasDefinition(alias)) {
      throw new Error(`Definition ${alias} already exists`)
    }

    this._definitions[alias] = Promise.all(deps).then((dependencies) => {
      return createDefinition(this._ceramic, definition).then(
        (doc) => ({ id: doc.id, dependencies }),
        (reason: any) => {
          delete this._definitions[alias]
          throw reason
        }
      )
    })
    return this._definitions[alias]
  }

  async addDefinition(definition: Definition, alias = definition.name): Promise<DocID> {
    const schemaAlias = this._schemaAliases[definition.schema]
    if (schemaAlias == null) {
      throw new Error('Schema for this definition has not been added')
    }
    const createdSchema = this._schemas[schemaAlias]
    if (createdSchema == null) {
      throw new Error('Schema alias for this definition could not be found')
    }

    const created = await this.createDefinition(alias, definition, [
      createdSchema.then(({ id }) => id),
    ])
    return created.id
  }

  hasTile(alias: string): boolean {
    return this._tiles[alias] != null
  }

  deleteTile(alias: string): boolean {
    if (this.hasTile(alias)) {
      delete this._tiles[alias]
      return true
    }
    return false
  }

  createTile<T extends Record<string, unknown>>(
    alias: string,
    contents: T,
    meta: Partial<DocMetadata> = {},
    deps: Array<Promise<DocID>> = []
  ): Promise<CreatedDoc> {
    if (this.hasTile(alias)) {
      throw new Error(`Tile ${alias} already exists`)
    }

    this._tiles[alias] = Promise.all(deps).then((dependencies) => {
      return createTile(this._ceramic, contents, meta).then(
        (doc) => ({ id: doc.id, dependencies }),
        (reason: any) => {
          delete this._tiles[alias]
          throw reason
        }
      )
    })
    return this._tiles[alias]
  }

  async addTile<T extends Record<string, unknown>>(
    alias: string,
    contents: T,
    meta: Partial<DocMetadata> = {}
  ): Promise<DocID> {
    const deps =
      meta.schema == null ? [] : [this._loadDoc(meta.schema).then((doc) => doc.versionId)]
    const created = await this.createTile(alias, contents, meta, deps)
    return created.id
  }

  async toPublished(): Promise<PublishedDocSet> {
    const definitions: Record<string, string> = {}
    const schemas: Record<string, string> = {}
    const tiles: Record<string, string> = {}

    const handleDefinitions = Object.entries(this._definitions).map(async ([alias, created]) => {
      const { id } = await created
      definitions[alias] = id.toString()
    })
    const handleSchemas = Object.entries(this._schemas).map(async ([alias, created]) => {
      const { id } = await created
      schemas[alias] = id.toUrl()
    })
    const handleTiles = Object.entries(this._tiles).map(async ([alias, created]) => {
      const { id } = await created
      tiles[alias] = id.toString()
    })
    await Promise.all([...handleDefinitions, ...handleSchemas, ...handleTiles])

    return { definitions, schemas, tiles }
  }

  async toSigned(): Promise<SignedDocSet> {
    const deps = new Set<string>()
    const docs: Record<string, Array<DagJWSResult>> = {}
    const definitions: Array<string> = []
    const schemas: Array<string> = []

    const addDoc = async (created: Promise<CreatedDoc>) => {
      const { id, dependencies } = await created
      dependencies.forEach((depid) => {
        deps.add(depid.toString())
      })
      const records = await this._ceramic.loadDocumentRecords(id)
      docs[id.toString()] = records.map((r) => r.value as DagJWSResult)
    }

    const handleDefinitions = Object.entries(this._definitions).map(async ([alias, created]) => {
      definitions.push(alias)
      return await addDoc(created)
    })
    const handleSchemas = Object.entries(this._schemas).map(async ([alias, created]) => {
      schemas.push(alias)
      await addDoc(created)
    })
    const handleTiles = Object.values(this._tiles).map(addDoc)
    await Promise.all([...handleDefinitions, ...handleSchemas, ...handleTiles])

    deps.forEach((id) => {
      if (docs[id] == null) {
        throw new Error(`Missing dependency: ${id}`)
      }
    })

    return { docs, definitions, schemas }
  }

  async toSignedJSON(): Promise<EncodedSignedDocSet> {
    const { docs, ...signed } = await this.toSigned()
    return { ...signed, docs: encodeSignedMap(docs) }
  }
}

export async function publishSignedDocSet(
  ceramic: CeramicApi,
  docSet: SignedDocSet
): Promise<void> {
  const schemas: Array<Promise<Doctype>> = []
  const others: Array<Promise<Doctype>> = []

  Object.entries(docSet.docs).forEach(([id, records]) => {
    const publish = publishRecords(ceramic, records)
    if (docSet.schemas.includes(id)) {
      schemas.push(publish)
    } else {
      others.push(publish)
    }
  })

  await Promise.all(schemas)
  await Promise.all(others)
}

export async function publishEncodedSignedDocSet(
  ceramic: CeramicApi,
  { docs, ...docSet }: EncodedSignedDocSet
): Promise<void> {
  return await publishSignedDocSet(ceramic, { ...docSet, docs: decodeSignedMap(docs) })
}
