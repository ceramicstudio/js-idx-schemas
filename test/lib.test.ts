import { CeramicApi } from '@ceramicnetwork/ceramic-common'

import { allSchemas, publishSchemas } from '../src'

declare global {
  const ceramic: CeramicApi
}

describe('lib', () => {
  test('publishSchemas', async () => {
    const schemas = Object.entries(allSchemas).map(([name, schema]) => ({ name, schema }))
    // eslint-disable-next-line
    const res = await publishSchemas({ ceramic, schemas })
    expect(res).toMatchSnapshot()
  })
})
