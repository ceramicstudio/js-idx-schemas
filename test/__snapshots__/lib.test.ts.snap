// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib publishSchemas 1`] = `
Object {
  "BasicProfile": "ceramic://bafyreibaqfnywxojlctrocccv5kmofgilpqhex4spfbp6sd3ortmrl3e2q",
  "DocIdDocIdMap": "ceramic://bafyreiecbrz74m4ptvpadfhtriuh5f2g5xe3pzhp5iugs4reiiw6bq5yxi",
  "DocIdMap": "ceramic://bafyreid7hbhnxcqyjs52grytmbsliaglvmkvdc3od3jfho34fp4f7akjsm",
  "StringMap": "ceramic://bafyreiagoofia7nmr4k6syboj6ckqft6utu2ne5k4d36vzh4udndglxety",
}
`;
